### PR TITLE
Support any attribute class in typed lists. Fixes #261

### LINKED ISSUE #261: Why MapAttribute is required for ListAttribute?
I think `ListAttribute` could get common`Attribute` like `UnicodeAttribute`
But it couldn't.

On pynamodb/attributes.py#L560-563
```python
        if of:
            if not issubclass(of, MapAttribute):
                raise ValueError("'of' must be subclass of MapAttribute")
            self.element_type = of
```

And so on, in this document section, http://pynamodb.readthedocs.io/en/latest/attributes.html#list-attributes

```python
from pynamodb.attributes import ListAttribute, NumberAttribute, UnicodeAttribute

class Office(Model):
    class Meta:
        table_name = 'OfficeModel'
    office_id = NumberAttribute(hash_key=True)
    employees = ListAttribute(of=UnicodeAttribute)
```

I wanna know what a right intent is. Thanks

### PR DESCRIPTION
**Pull request by jpinner-lyft**

### CODE REVIEW

**Comment by jpinner-lyft** [comment]
```
>>> from pynamodb.attributes import UnicodeAttribute, UnicodeSetAttribute, ListAttribute
>>> from pynamodb.models import Model
>>> class UnicodeSetListModel(Model):
...     foo = UnicodeAttribute(hash_key=True)
...     usl = ListAttribute(of=UnicodeSetAttribute)
... 
>>> uslm = UnicodeSetListModel()
>>> uslm.foo = 'foo'
>>> uslm.usl = [{'Hello'}, {'World'}, {'!'}]
>>> uslm._serialize()
{'attributes': {'usl': {'L': [{'SS': ['Hello']}, {'SS': ['World']}, {'SS': ['!']}]}}, 'HASH': 'foo'}
>>>
>>> uslm._deserialize({'usl': {'L': [{'SS': ['Hello']}, {'SS': ['World', '!']}]}})
>>> uslm.usl
[{'Hello'}, {'World', '!'}]
>>>
>>> uslm._deserialize({'usl': {'L': [{'SS': ['Hello']}, {'S': 'World!'}]}})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jpinner/workspace/PynamoDB/pynamodb/models.py", line 1128, in _deserialize
    value = attr.deserialize(value)
  File "/Users/jpinner/workspace/PynamoDB/pynamodb/attributes.py", line 975, in deserialize
    raise ValueError("Cannot deserialize elements of type: {}".format(attr_type))
ValueError: Cannot deserialize elements of type: S
```

File: pynamodb/attributes.py [hunk 1]
```diff
@@ -947,10 +947,16 @@ def serialize(self, values):
...
-            attr_class = (self.element_type()
-                          if self.element_type
-                          else _get_class_for_serialize(v))
-            rval.append({attr_class.attr_type: attr_class.serialize(v)})
+            attr_class = self._get_serialize_class(v)
+            if self.element_type and not isinstance(attr_class, self.element_type):
+                raise ValueError("List elements must be of type: {}".format(self.element_type.__name__))
+            attr_type = attr_class.attr_type
+            attr_value = attr_class.serialize(v)
+            if self.element_type and attr_value is None:
```
Comment (garrettheel) [inline pynamodb/attributes.py:955]: Why wouldn't we want to store `{"NULL": True}` if there's no explicit `element_type` specified?

File: pynamodb/attributes.py [hunk 2]
```diff
@@ -959,23 +965,43 @@ def deserialize(self, values):
...
-            attr_value = _get_value_for_deserialize(v)
-            deserialized_lst.append(class_for_deserialize.deserialize(attr_value))
+            attr_type, attr_value = next(iter(v.items()))
+            attr_class = self._get_deserialize_class(attr_type)
+            if self.element_type and attr_type == NULL:
+                # When attribute values serialize to "None" (e.g. empty sets) we store {"NULL": True} in DynamoDB.
+                attr_type = self.element_type.attr_type
+                attr_value = None
+            if self.element_type and self.element_type.attr_type != attr_type:
+                raise ValueError("Cannot deserialize elements of type: {}".format(attr_type))
```
Comment (garrettheel) [inline pynamodb/attributes.py:975]: Can we add the expected `attr_type` to this message too? Just to help folks understand the issue more quickly

**Review by garrettheel — COMMENTED** [review]
Generally looks good so far. Looking forward to getting this in.

File: pynamodb/attributes.py (hunk 1 above)
Comment (jpinner-lyft) [reply pynamodb/attributes.py]: updated

File: pynamodb/attributes.py (hunk 2 above)
Comment (jpinner-lyft) [reply pynamodb/attributes.py]: updated with attribute class name

**Review by garrettheel — APPROVED** [review]
👍

**Pull request merged**